evm/src/invariant_fuzz.rs

```diff
@@ -0,0 +1,363 @@
+                                    }),
+                                );
+                                break 'all
+                            } else {
+                                // This will panic and get caught by the executor
+                                if !executor.is_success(
+                                    invariant_address,
+                                    reverted,
+                                    state_changeset.expect("we should have a state changeset"),
+                                    false,
+                                ) {
+                                    invariant_doesnt_hold.borrow_mut().insert(
+                                        func.name.clone(),
+                                        Some(InvariantFuzzError {
+                                            test_error: proptest::test_runner::TestError::Fail(
+                                                format!(
+                                                    "{}, reason: '{}'",
+                                                    func.name,
+                                                    match foundry_utils::decode_revert(
+                                                        result.as_ref(),
+                                                        abi
+                                                    ) {
+                                                        Ok(e) => e,
+                                                        Err(e) => e.to_string(),
+                                                    }
+                                                )
+                                                .into(),
+                                                inputs.clone(),
+                                            ),
+                                            return_reason: "".into(),
+                                            revert_reason: foundry_utils::decode_revert(
+                                                result.as_ref(),
+                                                abi,
+                                            )
+                                            .unwrap_or_default(),
+                                            addr: invariant_address,
+                                            func: func.short_signature().into(),
+                                        }),
+                                    );
+                                    break 'all
+                                }
+                            }
+                        }
+                        // push test case to the case set
+                        fuzz_cases.borrow_mut().push(FuzzCase {
+                            calldata: calldata.clone(),
+                            gas,
+                            stipend,
+                        });
+                    } else {
+                        // call failed, continue on
```

Comment:
Say the 3rd call in a sequence reverts (i.e. the fuzz call itself reverts, not a failed invariant), do we generate a new call or move on to the fourth call in the sequence? Looks like the latter, so wondering if we should have a flag to instead generate a new call, to help mitigate situations where many calls in the sequence revert?

For reference, relevant portion of dapptools docs:

> Note that a revert in any of the randomly generated call will not trigger a test failure. The goal of invariant tests is to find a state change that results in a violation of the assertions defined in the body of the test method, and since reverts do not result in a state change, they can be safely ignored. Reverts within the body of the invariant* test method will however still cause a test failure.

Reply:
> although having a report on the number of reverted calls might be interesting, so the user can adjust his settings accordingly.

Good points, this seems like a good compromise. Have the option to report min/max/mean/median number of reverts. Another idea is to have a `targetSelectors()` method which returns a tuple of `(address, bytes4[])` so you can have fine grained control over the allowed methods (echidna has similar functionality where you can allow/block function selectors)

Reply:
Ooooh I like the idea of `targetSelectors()`